graphqlapi/graphiql/graphiql.go

```diff
@@ -0,0 +1,224 @@
+// Based on `graphiql.go` from https://github.com/graphql-go/handler
+// only made RenderGraphiQL a public function.
+package graphiql
+
+import (
+	"encoding/json"
+	"html/template"
+	"net/http"
+	"strings"
+)
+
+// graphiqlData is the page data structure of the rendered GraphiQL page
+type graphiqlData struct {
+	GraphiqlVersion string
+	QueryString     string
+	Variables       string
+	OperationName   string
+	AuthKey         string
+	AuthToken       string
+}
+
+func AddMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if strings.HasPrefix(r.URL.Path, "/weaviate/v1/graphql") && r.Method == http.MethodGet {
+			renderGraphiQL(w, r)
+		} else {
+			next.ServeHTTP(w, r)
+		}
+	})
+}
+
+// renderGraphiQL renders the GraphiQL GUI
+func renderGraphiQL(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("WWW-Authenticate", `Basic realm="Provide your key and token (as username as password respectively)"`)
+
+	user, password, authOk := r.BasicAuth()
+	if !authOk {
+		http.Error(w, "Not authorized", 401)
+		return
+	}
+
+	query_params := r.URL.Query()
+
+	t := template.New("GraphiQL")
+	t, err := t.Parse(graphiqlTemplate)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Attempt to deserialize the 'variables' query key to something reasonable.
+	var query_vars interface{}
+	err = json.Unmarshal([]byte(query_params.Get("variables")), &query_vars)
+
+	var varsString string
+	if err == nil {
+		vars, err := json.MarshalIndent(query_vars, "", "  ")
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		varsString = string(vars)
+		if varsString == "null" {
+			varsString = ""
+		}
+	}
+
+	// Create result string
+	d := graphiqlData{
+		GraphiqlVersion: graphiqlVersion,
+		QueryString:     query_params.Get("query"),
+		Variables:       varsString,
+		OperationName:   query_params.Get("operationName"),
+		AuthKey:         user,
+		AuthToken:       password,
+	}
+	err = t.ExecuteTemplate(w, "index", d)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+
+	return
+}
+
+// graphiqlVersion is the current version of GraphiQL
+const graphiqlVersion = "0.11.11"
```

Comment:
Probably should move this constant to the top of the file, below the imports, above the struct definition.